Bid schema, validation, & current price calculation

rstein66 opened the pull request:
- Seller should not be able to place a bid on their own item  
- Buyer can place a bid IFF (1) the amount is higher than the current price and (2) the auction is open
- Buyer can place back-to-back bids

> Apologies for not noting & cleaning up the messy history prior to creating a PR!

rstein66 committed:
Create bids migration

rstein66 committed:
Create bids migration

rstein66 committed:
Create bids migration

rstein66 committed:
Display countdown to auction end using LiveView

rstein66 committed:
Include attribution

rstein66 committed:
Fix typo

rstein66 committed:
Rebase `bid` on top of view auction details (PR #6)

xray committed:
remove package-lock.json

xray committed:
Merge branch 'bid' of https://github.com/xray/Renaissance into bid

rstein66 committed:
Extract adapt & valid helpers

rstein66 committed:
Fix test file

rstein66 committed:
WIP- add backend support for place bid

rstein66 committed:
Add travis CI config file

rstein66 committed:
Set exclusive lower bound for bid amount

Enforce business constraint that amount needs to be
- Positive
- Greater than the auction item's seller-specified
starting price
- Greater than the current highest bid on the item

rstein66 committed:
Merge branch 'bid-backend' into bid

rstein66 committed:
Ensure bidder & seller are not the same user

Also, clean up–more accessible (for front-end, etc)
exclusive lower bound getter

rstein66 committed:
Merge branch 'master' into bid-backend

rstein66 committed:
Accept bids only while auction is open

rstein66 committed:
Include codecov dep

rstein66 committed:
Update Travis config

rstein66 committed:
Fix Travis config

rstein66 committed:
Remove unused deps

rstein66 committed:
Test that user can place back-to-back bids

rstein66 committed:
Remove branch-specific Travis CI badge

rstein66 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Bid schema, validation, and current price calculation](https://trello.com/c/mpeuSx33/15-bid-schema-validation-and-current-price-calculation)

rstein66 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Setup Continuous Integration (CI) to run tests.](https://trello.com/c/oEIjKPC3/12-setup-continuous-integration-ci-to-run-tests)

kevbuchanan commented on lib/renaissance/auctions/auction.ex:
```diff
@@ -21,24 +23,15 @@ defmodule Renaissance.Auction do
-    |> validate_date()
+    |> foreign_key_constraint(:seller_id, name: :auctions_user_id_fkey)
+    |> Constraint.amount_constraint(:price)
+    |> end_datetime_constraint()
```
`_constraint` implies something being applied in the database that Ecto should know about. This should be `validate_end_datetime`

ssunday replied:
+1

kevbuchanan commented on lib/renaissance/bids/bids.ex:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def get_highest_bid(auction_id) do
+    query = "SELECT bidder_id, amount, created_at
```
This query should be constructed through Ecto: https://hexdocs.pm/ecto/Ecto.Query.html

rstein66 replied:
Resolved in 845f45e.

kevbuchanan commented on lib/renaissance/helpers/adapt.ex:
```diff
@@ -0,0 +1,19 @@
+defmodule Renaissance.Helpers.Adapt do
```
The `Adapt` name is a little bit unclear. Maybe just `Renaissance.Format.Money`? Everything in here has to do with Money.

ssunday commented on .gitignore:
```diff
@@ -245,3 +245,4 @@ bh_unicode_properties.cache
 GitHub.sublime-settings
 
 # End of https://www.gitignore.io/api/sublimetext
+assets/package-lock.json
```
Why remove package-lock.json from VCS?

xray replied:
I found it to be adding a lot of noise to diffs, but I'll just commit it separately instead.

kevbuchanan commented on lib/renaissance/helpers/compare.ex:
```diff
@@ -0,0 +1,33 @@
+    end
+  end
+
+  defp return_money_max(m) do
```
`to_money` might be a better name.

ssunday commented on lib/renaissance/auctions/auctions.ex:
```diff
@@ -1,50 +1,51 @@
   end
 
+  def get_seller_id(id) do
+    if exists?(id), do: get!(id).seller_id
```
I'm not 100% on the ecto optimization of it, but I'm guessing ecto pulls out the full record with `get!` in which case why have methods in the repo to pull specific things if it's not optimizing the query? Why not have a find_if_exists type call that does get and then in the methods get the attributes?

rstein66 replied:
I wouldn't be at all surprised if that's true & can see the value, but I don't know if it would be worthwhile at present. As I see it, at present, this is the only function that retrieved a record, sans preloading, & extracts a specific field without adding additional functionality.

ssunday commented on lib/renaissance/auctions/auctions.ex:
```diff
@@ -1,50 +1,51 @@
+  def get_current_amount(id) do
+    larger =
+      Bids.get_highest_bid_amount(id)
+      |> Compare.money_max(get_starting_amount(id))
```
would a case piped into this be cleaner?

ssunday commented on lib/renaissance/bids/bids.ex:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def get_highest_bid(auction_id) do
+    query = "SELECT bidder_id, amount, created_at
```
Does this have to be in raw SQL?

rstein66 replied:
Resolved in 845f45e.

ssunday commented on lib/renaissance/bids/bids.ex:
```diff
@@ -0,0 +1,53 @@
+  alias Renaissance.Helpers.Adapt
+
+  def place_bid(params) do
+    params = Adapt.format_amount(params, "amount")
```
so this is basically creating it...do we need to have it use a special name instead of the standard ones?

rstein66 replied:
Resolved in 15d9389.

ssunday commented on lib/renaissance/helpers/compare.ex:
```diff
@@ -0,0 +1,33 @@
+
+  def amount?(%Money{}), do: true
+
+  def amount?(p) do
```
can't you just return `is_integer(p)`?

ssunday commented on lib/renaissance/helpers/compare.ex:
```diff
@@ -0,0 +1,33 @@
+  def money_max(m1, m2) do
+    cond do
+      amount?(m1) && amount?(m2) == true ->
+        if Adapt.money_value(m1) >= Adapt.money_value(m2) do
```
could this be extracted out, it's hard to read

kevbuchanan replied:
Can you use the `Money` comparison functions here? https://hexdocs.pm/money/Money.html#compare/2

ssunday commented on lib/renaissance/helpers/constraint.ex:
```diff
@@ -0,0 +1,42 @@
+  def open_constraint(changeset, field) do
+    auction_id = get_change(changeset, :auction_id, 0)
+
+    if Auctions.open?(auction_id) in [nil, true] do
```
If the record doesn't exist....shouldn't that be a diff error?

kevbuchanan commented on priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs:
```diff
@@ -0,0 +1,19 @@
+  use Ecto.Migration
+
+  def up do
+    rename table(:bids), :inserted_at, to: :created_at
```
I'm curious why it was necessary to change the conventional `inserted_at` and `updated_at` for bids...?

ssunday replied:
^^

rstein66 replied:
I will remove this. 

_[I had changed it because, in my mind, having only one timestamp–`created_at`–makes it clearer that the entries in `bids` will not be modified, which contrasts with the entries in `users` & `auctions`.]_

kevbuchanan replied:
I think that's fine and makes sense.

kevbuchanan commented on test/renaissance/users/users_test.exs:
```diff
@@ -1,40 +1,49 @@
 
-      assert user.email == @valid_attrs.email
-      assert user.password_hash != @valid_attrs.password
+      assert Repo.exists?(User) == true
```
`Repo.exists?(User)` will test that there is any user in the users table. I'm not sure if that's what's intended here. If so, a more clear and specific assertion would be nice.

ssunday commented on test/renaissance/auctions/auctions_test.exs:
```diff
@@ -17,66 +17,79 @@ defmodule Renaissance.Test.AuctionsTest do
-          Auctions.create_auction(invalid_seller_id, @auction_two)
-        end
+    test "false when no auction with given id" do
+      assert Auctions.exists?(0) == false
```
`asset_false` works, no?

rstein66 replied:
Resolved with 46d8ca2 & 209badd.

_(Personally, I find [`refute`](https://hexdocs.pm/ex_unit/ExUnit.Assertions.html#functions) less clear, especially because it checks if the value is `in [nil, false]`, but that's just me.)_

kevbuchanan commented on lib/renaissance/auctions/auctions.ex:
```diff
@@ -1,50 +1,51 @@
   end
 
+  def get_seller_id(id) do
+    if exists?(id), do: get!(id).seller_id
```
There seems to be this pattern of `exists?(thing), do: the_operation(thing)`, which isn't really necessary, since you can just handle nil from get:

```
case get(id) do
  nil -> nil
  auction -> auction.seller_id
end
```

ssunday replied:
IMO, should this even be checking it? Shouldn't checking for record existing be higher up the handling order?

rstein66 replied:
Where do you mean by higher up? The form?

If yes, the form errors (beyond those that can be checked using HTML tags) are generated by the changeset validation. I'm unaware of how else to generate them. All the examples I've seen generate form errors by means of the changeset validation; that said, I have not researched this topic in any depth.

kevbuchanan commented on lib/renaissance/helpers/constraint.ex:
```diff
@@ -0,0 +1,42 @@
+  alias Renaissance.Auctions
+
+  def amount_constraint(changeset, field) do
+    auction_id = get_change(changeset, :auction_id, 0)
```
How does this work from Auction, since Auction doesn't have an auction_id?

kevbuchanan commented on lib/renaissance/helpers/constraint.ex:
```diff
@@ -0,0 +1,42 @@
+    proposed = get_change(changeset, field, 0) |> Adapt.money_value()
+    current = Auctions.get_current_amount(auction_id) |> Adapt.money_value()
+
+    if proposed > current do
```
This check is not guaranteed to be accurate with concurrent bids being placed. Is that an issue? And if so, how could you ensure that a valid bid must always be placed higher than the current highest bid?

rstein66 replied:
I don't think it's an issue, at least based on my current understanding, which is as follows:
- If there are concurrent [valid] bids of difference amounts, then (for all practical purposes) it's as if the users' with the lower bids were out-bid, and therefore no special handling is needed. 
-  I updated `get_highest_bid` ([updated query](https://github.com/xray/Renaissance/blob/845f45ee2e/lib/renaissance/bids/bids.ex#L30);  [test](https://github.com/xray/Renaissance/blob/845f45ee2e/test/renaissance/bids/bids_test.exs#L117)) to return the earliest (or at least first recorded) bid if there are multiple valid bids with the highest amount for a given auction so that, if the auction closes in this state, the winner will be the first person to place the bid.

kevbuchanan commented on lib/renaissance/helpers/constraint.ex:
```diff
@@ -0,0 +1,42 @@
+
+  def bidder_constraint(changeset, field) do
+    bidder_id = get_change(changeset, field)
+    auction_id = get_change(changeset, :auction_id, 0)
```
Why do we default to 0 here?

kevbuchanan commented on lib/renaissance/helpers/constraint.ex:
```diff
@@ -0,0 +1,42 @@
+    auction_id = get_change(changeset, :auction_id, 0)
+    seller_id = Auctions.get_seller_id(auction_id)
+
+    if is_nil(seller_id) or bidder_id != seller_id do
```
This condition is a little bit hard to parse. I guess I would expect a little bit more clarity on the condition that we actually care about here:

```
case {seller_id, bidder_id} do
  {nil, _} -> changeset # How do we even end up in this place?
  {seller_id, bidder_id} when seller_id == bidder_id -> add_error(changeset, "can't bid on the item you're selling")
  _ -> changeset
end
```

kevbuchanan commented on lib/renaissance/helpers/constraint.ex:
```diff
@@ -0,0 +1,42 @@
+  def open_constraint(changeset, field) do
+    auction_id = get_change(changeset, :auction_id, 0)
+
+    if Auctions.open?(auction_id) in [nil, true] do
```
The fact that `open?` returns nil is a bit weird. `something?` should only return true or false.

ssunday replied:
agreed

rstein66 committed:
Begin clean up of validators

rstein66 committed:
Replace `assert ... == false` tests with `refute`

rstein66 committed:
Replace domain-specific method names with standard

rstein66 committed:
Refactor excl. lower bound implementation

Clean up & simplify the logic behind setting an exclusive lower
bound for a given bid (functionality added in 593eef8)

rstein66 committed:
Merge branch 'master' into bid-backend

rstein66 committed:
Resolve merge conflicts

xray commented:
![](https://github.trello.services/images/mini-trello-icon.png) [User can bid on an item. (8.25pts)](https://trello.com/c/3N9qLZTl/9-user-can-bid-on-an-item-825pts)

xray commented:
![](https://github.trello.services/images/mini-trello-icon.png) [User can bid on an item. (8.25pts)](https://trello.com/c/3N9qLZTl/9-user-can-bid-on-an-item-825pts)
